jep/0000/README.adoc

```diff
@@ -0,0 +1,238 @@
+
+=== Target Audience
+
+We have identified two personas as target Audience :
+
+*Ben* is system administrator in the company, managing services across all teams : LDAP, Mail, but also developers ressources like 
+repositories and CI servers. Ben’s favorite IDE is vi, he’s preferred language is bash (or maybe ruby|python) and only knows Java for the 
+amount of memory it requires and the delay to start a JVM. So he’s not a big fan of writing groovy scripts. He mostly knows Jenkins by the 
+UI and used to backup for whole JENKINS_HOME folder.
+
+*Seb* is working in a startup as DevOps engineer. He’s full time working with product team as a developer and managing development process 
+and resources (which makes 2 full times for one guy. This is a startup I said). Seb has deployed kubernetes as general purpose 
+infrastructure on the Cloud and deployed a git repo and few other services for his team. A Jenkins master is running there as well. He use 
+to upgrade jenkins by running latest official docker image and re-using previous jenkins-home volume.
+
+== Reasoning
+
+Configuration as Code as a simple text file with both documentation and schema would make it possible for any Jenkins user to replicate 
+the configuration he use to setup by hand on web UI.
+
+This is a major differentiator vs Groovy init scripts used by many Advanced Jenkins users, who are confident with internal APIs and Groovy 
+syntax. Using a basic text file format with validation make this feature available to arbitrary DevOps team without need to be familiar 
+with Jenkins or Groovy.
+
+=== File format
+
+To avoid Configuration as Code to be tied to some development community, we selected YAML as format to define Jenkins configuration.
+
+YAML allows to :
+
+* Have a plain text, human readable format
+* Include comments to provide runnable sample configuration files
+* Be language ecosystem agnostic
+* Support JSON-schema validation
+
+=== Configuration mechanism
+
+Configuration file define a tree model. For every node we need to find the matching Jenkins component. Relying on @Symbol annotation is an 
+efficient way to identify components with a human friendly short name. For plugin which didn’t (yet) adopted this annotation, we can rely 
+on some convention. A common pattern is to name an implementation class as prefix + API class name, like “LDAPSecurityRealm”. As we know 
+the API we are looking for implementation, we can establish a natural short name for this implementation as “ldap” and offer a Symbol-like 
+short name to end user for his configuration file.
+
+
++DataBoundSetters+ & +DataBoundConstructors+ offer a natural way to construct Jenkins components from a set of key=value pairs. Most 
+jenkins component (*) do rely on them and as such offer a 1:1 match between internal data structure and web UI configuration forms. 
+
+(*) We noticed many +Descriptor+s do rely on manual parsing of +JSONObject+. We will need to fix them 
```

Comment:
We probably need to fix them independently of this JEP. The problem is that there are complex structures and custom forms sometimes. Would be great to know a list of plugins and core changes required for MVP.

Reply:
> will need to fix them

This seems to be a *massive* change -- I wouldn't be surprised if this affected _hundreds_ of plugins, not to mention any plugins that would be newly incompatible since `newInstance` isn't `final` (and probably never will be).

I would strongly suggest looking into a way to get this implemented without requiring all `@DataboundWhatever` for better compatibility.

Reply:
If the plugin offers extension points for such custom cases, it would be already helpful.

Reply:
if an extension point is introduce and plugins have to adopt them, we are back in the issue to require hundreds plugins to be fixed. `@DataboundSetter` should be enough. Main issue AFAICT is one can't reset to null a Databound attribute.

Reply:
> we are back in the issue to require hundreds plugins to be fixed

I recommend you look into how many plugins use non-databound form submission data. I don't see how requiring moving to `@DataboundX` is different.

Reply:
`@DataboundSetter` is obvious way to document attributes and types for form submission. With custom databinding we can't guess this model, and generate a configuration-as-code schema/doc. 

I've started reviewing existing Descriptors and indeed there's various patterns. optional blocks typically introduce a boolean pseudo-attribute and nested JSON document, while Descriptor use nullable attributes. This hardly can be translated into a `@DataBoundSetter` (won't be invoked on null value)

Reply:
Why not both? Supporting pure DataBound plugins by a generic solution is perfectly fine. Regarding else, there should extension points which allow doing it without reworking to DataBounds. I think that the latter thing is inevitable, migration to databound forms may be even impossible in some cases

Reply:
prototype already defines a `Configurator` extension point plugins can implement. But then they'll depend on this plugin. Issue I can see doing this is plugins will need to implement configuration twice : one for c-as-c and the other for web UI binding. Using DataBound mechanism ensure we are un sync.
I wonder what would prevent use of databound. After all it's just a replacement from `foo = json.getXX()` to `@DataBoundSetter setFoo(xx)`.  If the raw form data is then computed to something else, same code can move into such a setter.